rust/src/dcerpc/dcerpc.rs

```diff
@@ -0,0 +1,620 @@
+        }
+        // Shouldn't happen
+        return 0;
+    }
+
+    pub fn parse_fragment_data(&mut self, input: &[u8], input_len: u16) -> u16 {
+        let mut retval: u16 = 0;
+        let hdrflags1 = self.get_hdr_flags1();
+        let fraglenleft = self.fraglenleft;
+
+        // Update the stub params based on the packaet type
+        match self.get_hdr_pkt_type() {
+            DCERPC_TYPE_REQUEST => {
+                if let Some(ref mut req) = self.dcerpcudprequest {
+                    retval = evaluate_stub_params(
+                        input,
+                        input_len,
+                        hdrflags1,
+                        fraglenleft,
+                        &mut req.stub_data_buffer,
+                        &mut req.stub_data_buffer_len,
+                    );
+                }
+            }
+            DCERPC_TYPE_RESPONSE => {
+                if let Some(ref mut resp) = self.dcerpcudpresponse {
+                    retval = evaluate_stub_params(
+                        input,
+                        input_len,
+                        hdrflags1,
+                        fraglenleft,
+                        &mut resp.stub_data_buffer,
+                        &mut resp.stub_data_buffer_len,
+                    );
+                }
+            }
+            _ => {
+                SCLogDebug!("Unrecognized packet type");
+                return 0;
+            }
+        }
+        // Update the remaining fragment length
+        self.fraglenleft -= retval;
+
+        retval
+    }
+
+    pub fn parse_dcerpc_udp_header(&mut self, input: &[u8]) -> i32 {
+        match parser::dcerpc_parse_header(input) {
+            Ok((_leftover_bytes, header)) => {
+                if header.rpc_vers != 4 {
+                    SCLogDebug!("DCERPC UDP Header did not validate.");
+                    return -1;
+                }
+                let mut uuidentry = DCERPCUuidEntry::new();
+                let auuid = header.activityuuid.to_vec();
+                uuidentry.uuid = auuid;
+                self.uuid_list.push(uuidentry);
+                self.dcerpchdrudp = Some(header);
+                80
+            }
+            Err(nom::Err::Incomplete(_)) => {
+                // Insufficient data.
+                SCLogDebug!("Insufficient data while parsing DCERPC request");
+                -1
+            }
+            Err(_) => {
+                // Error, probably malformed data.
+                SCLogDebug!("An error occurred while parsing DCERPC request");
+                -1
+            }
+        }
+    }
+
+    pub fn parse_dcerpc_udp(&mut self, input: &[u8]) -> i32 {
+        // Call header parser first
+        let mut parsed = self.parse_dcerpc_udp_header(input);
+        if parsed == -1 {
+            return -1;
+        }
+
+        // Input left after successful header parsing should be 80 less
+        // than the original length
+        let mut input_left = input.len() as i32 - 80;
```

Comment:
could this underflow is `input.len()` is smaller than 80? I see no explicit checks for this.

Reply:
hmm maybe. My instinct was that input len should at least be 80 if the code reaches here else it should return -1 as header parsing will fail because of incomplete header (less than 80). What do you think?

Reply:
probably best to be explicit about it. It will be more robust especially against future changes.